### PR TITLE
fix(json): report a malformed JSON array element instead of dropping the rest of the file (fixes #12755)

### LINKED ISSUE #12755: A malformed JSON array element is treated as "need more data", so the rest of the file is silently dropped
## Describe the bug

`ArrayToNdjsonPush` — the push-based JSON-array-to-NDJSON adapter behind
`SpiceJsonDecoder`, used for every `json_format: array` dataset
(`crates/dataformat-json/src/source.rs:505`, `matches!(format, Format::Array)`) —
treats a **genuine JSON syntax error as "not enough data yet"**. The malformed
element and every element after it are dropped, no error is raised, and the query
returns a short result.

`crates/dataformat-json/src/stream.rs:1030-1039`:

```rust
Some(Err(e)) => {
    // Check if this is a "need more data" error
    if e.classify() == Category::Eof || e.classify() == Category::Syntax {
        // This is expected when we have partial data - just wait for more
        return Ok(());
    }

    // This is a real syntax error
    return Err(io::Error::new(io::ErrorKind::InvalidData, e));
}
```

`serde_json` already separates the two cases exactly:

| input | classify() |
|---|---|
| `{"a": 1` (truncated) | `Eof` |
| `{"a": "abc` (truncated) | `Eof` |
| `{"a" 1}` (missing colon) | `Syntax` |
| `{"a": tru3}` (bad token) | `Syntax` |
| `{"a": 1,}` (trailing comma) | `Syntax` |

So `Category::Eof` is the whole "need more data" case, and adding `Category::Syntax`
to it swallows every real error. The comment on the line below — *"This is a real
syntax error"* — guards a branch that a syntax error can no longer reach.

## To Reproduce

```rust
let mut adapter = ArrayToNdjsonPush::new();
adapter.push_bytes(br#"[{"a":1},{"a" 2},{"a":3}]"#).unwrap();  // Ok(())  <-- no error
assert_eq!(adapter.try_read(), ReadResult::Ready(b"{\"a\":1}\n".to_vec()));
assert_eq!(adapter.try_read(), ReadResult::NotReady);           // {"a":3} never arrives
assert!(!adapter.is_complete());                                // and never will
```

Observed: one row. Expected: an error naming the malformed element.

`{"a":3}` is well-formed and follows the malformed element, but the parser never
advances past the failure — every later `push_bytes` re-runs `process_buffer`,
re-parses the same bytes from the same offset, and fails the same way. So the
buffer also grows for the rest of the file while nothing is emitted.

Through the decoder this is silent: `SpiceJsonDecoder::decode`
(`file_format.rs:485`) returns `Ok(buf_len)` — "consumed everything" — regardless,
so the scan ends normally with the rows missing.

## Second instance of the same shape, same function

`ParsingState::ExpectingArrayStart` (`stream.rs:973-984`) collapses the same two
cases:

```rust
if matches!(skip_ws_until(&mut cursor, b'['), Ok(())) { ... } else {
    // Not enough data yet
    return Ok(());
}
```

`skip_ws_until` already distinguishes them by `ErrorKind` — `UnexpectedEof` for a
buffer that has not reached the `[` yet, `InvalidData` for input that is not a JSON
array at all — and only the first is "not enough data yet". A `json_format: array`
dataset pointed at a JSONL or object payload therefore reads as an empty result
rather than a configuration error.

## Expected behavior

A malformed JSON array reports an error naming what was wrong. Only a genuinely
incomplete buffer waits for more bytes.

## Impact

Silent row loss on the `json_format: array` read path: a single malformed element
truncates the dataset at that point with no error, warning, or short-read signal.
The equivalent pull-based path (`ArrayToNdjson::fill_pending`, `stream.rs:146-158`)
gets this right — it only accepts `Category::Syntax` after peeking and confirming a
`]`, i.e. the empty-array case, and returns the error otherwise. The push path does
not need that allowance at all, because it handles the empty array structurally in
`ExpectingFirstElement` before it ever deserializes.

### LINKED ISSUE #12780: A truncated or empty JSON array body succeeds instead of erroring
### Describe the bug

`ArrayToNdjsonPush` has no end-of-input transition, so a JSON array body that the stream never
finishes delivering succeeds as a short (or empty) result instead of erroring.

`push_bytes` treats "the buffer stops part-way through" as *wait for more bytes* and returns
`Ok(())`. That is correct mid-stream, but nothing ever tells the parser the stream ended, so the
last such `Ok(())` is final. `SpiceJsonDecoder::flush`
(`crates/dataformat-json/src/file_format.rs:504`) only flushes the inner decoder:

```rust
fn flush(&mut self) -> Result<Option<RecordBatch>, ArrowError> {
    let projected_schema = Arc::clone(&self.projected_schema);
    self.inner.flush().map(move |batch| { /* … */ })
}
```

`ArrayToNdjsonPush::is_complete()` exists but no caller consults it at end of input, so the
`ParsingState` the stream actually stopped in is never checked.

Affected bodies, all of which currently succeed:

- an empty or whitespace-only payload — never leaves `ExpectingArrayStart`, so it reads as a
  successful zero-row result
- a truncated array (`[{"a":1},{"a":2` , connection dropped) — the rows parsed so far are
  returned and the unterminated tail is dropped silently

Both are indistinguishable from a legitimately empty or complete array, so a truncated read
becomes silent data loss rather than a failed query.

### Expected behavior

End of input should be an explicit operation that rejects any state other than
`ParsingState::Complete`, wired into the decoder's `flush`, so an unterminated array body fails
loudly instead of returning a partial result.

### Additional context

Pre-existing — trunk's `push_bytes` returns `Ok(())` on the same paths, and neither version has an
end-of-input transition. Raised by Copilot in review of #12777, which is scoped to a *different*
part of the class (a malformed or buffer-cut-short element mid-array) and does not touch `flush` or
add an end-of-input path. Filed separately rather than widening that PR, since the fix adds API
surface (a finalize/finish on the pusher plus a decoder wiring change) and wants its own tests.

Thread: https://github.com/spiceai/spiceai/pull/12777#discussion_r3739605348

### PR DESCRIPTION
**Pull request by claudespice**
## Summary

`ArrayToNdjsonPush` — the push-based JSON-array-to-NDJSON adapter behind every `json_format: array` dataset — treated a genuine JSON syntax error as "not enough data yet":

```rust
if e.classify() == Category::Eof || e.classify() == Category::Syntax {
    return Ok(());   // "just wait for more"
}
```

`serde_json` separates the two cases: a buffer that stops part-way through an element classifies as `Category::Eof`, and `Category::Syntax` is what it reports for input no further bytes can repair. Accepting both meant the malformed element **and every element after it** were dropped: each later push re-parsed the same bytes from the same offset and failed the same way, so nothing was emitted again while the buffer grew for the rest of the file. `push_bytes` returned `Ok`, `SpiceJsonDecoder::decode` reported the whole buffer consumed, and the scan ended normally with the rows missing.

Fixing that surfaced two more silent-wrongness cases on the same buffer boundary, both fixed here rather than left for follow-ups.

## Changes

**1. A malformed element is reported, not swallowed.** Only `Category::Eof` waits.

**2. A syntax error that is really a truncation still waits.** `Category::Eof` does not cover every incomplete buffer: one that stops inside a numeric literal — `-`, `1.`, `1e`, `1e-` — is reported as `Category::Syntax`, because what serde has read is not a number and it cannot know more is coming. Reporting every syntax error would have failed a *valid* array whose chunk boundary landed mid-number. A syntax error now counts as malformed only once the buffer ends on a byte no continuation could extend into a value, so the malformation still surfaces as soon as the element completes.

**3. A scalar cut short by the buffer is no longer emitted as its prefix.** A scalar has no closing delimiter, so serde ends it at the end of the buffer and reports *success*: `[123` + `456]` emitted the row `123` — a value the file never contained — and then failed on the leftover `456`. Objects, arrays and strings all end on a delimiter serde requires, so only a scalar running to the buffer's end is ambiguous.

**4. An array the input ended in the middle of is reported.** The adapter had no end-of-input transition at all, so a file cut short mid-array simply stopped producing rows: `[{"a":1},{"a":2}` emitted both rows, left `is_complete()` false, and returned no error. A file truncated by a partial fetch or an interrupted upload read as a complete, shorter table — and change 3 above widened that window, since the trailing scalar of `[1,2,3` is now held back waiting for bytes that never arrive. `ArrayToNdjsonPush::finish` rejects every state but `Complete`, and `SpiceJsonDecoder::flush` calls it. `flush` is the only correct hook: `can_flush_early()` is false and `DecoderDeserializer::next` flushes solely when `decode` returns 0, which happens only for the empty sentinel pushed once the input is exhausted, so it cannot fire mid-stream. A valid file cannot be mistaken for a truncated one — `Format::Array` is in `needs_non_repartitioned`, so files are never range-split into fragments, and the decoder is built inside the per-file `open()` future, so state never carries across files.

**5. A body that is not a JSON array is reported.** `ParsingState::ExpectingArrayStart` collapsed `skip_ws_until`'s two outcomes — `UnexpectedEof` (the `[` has not arrived yet) and `InvalidData` (the body is not an array at all). A `json_format: array` dataset pointed at an NDJSON or object payload read as an empty result rather than a configuration error.

**6. Content after the closing `]` is reported.** Reaching `]` ends the array but not the input: the closing bracket was drained and anything after it stayed in the buffer unparsed, while every later push was accumulated and ignored. A body holding more than the single array it was read as — a second concatenated array, or a larger document that merely starts with one — yielded that first array's rows and reported success, which is the same silent short read as change 4. `[1]garbage`, `[1][2]` and `[{"a":1}] {"b":2}` all emitted their first array's rows with `is_complete()` true and `finish()` returning `Ok`. The check runs at all three points where the state can be or become `Complete`, including the early return that swallowed later pushes, so trailing bytes are caught whether they share the push that closed the array or arrive after it. Whitespace is still accepted, so a file ending in a newline reads as it always did.

**7. The verified whitespace tail is dropped rather than rescanned.** Bytes arriving after the array closes are appended to the buffer but never parsed, so checking them on every push re-scanned the accumulated tail each time. Whitespace that has already been checked cannot become anything else, so it is discarded once verified — which also stops the tail being retained for the rest of the scan.

The buffered reader already rejected every one of these bodies, so the two paths over the same file disagreed — `[{"a":1},{"a":2}` was `Err(UnexpectedEof)` there and two silent rows here; `[1,2,3` was `Err(InvalidData)` there and two silent rows here. They now agree.

The pull-based path (`ArrayToNdjson::fill_pending`) was checked and is correct: its empty-array allowance peeks from just after the last parsed element, so a real syntax error does not reach it. No other site in the tree treats a parse error as "need more data" (`rg 'classify\(\)|Category::'`).

## Behavior change

Intentional, and the point of the fix: a `json_format: array` dataset with a malformed element **now fails loudly instead of returning a short result**, and one pointed at a non-array body now reports a configuration error instead of reading as empty. A truncated file now fails instead of returning the rows that happened to arrive. All three previously produced silently wrong data.

A body that holds more than one JSON array, or any non-whitespace content after the closing `]`, now fails instead of silently returning only its first array.

One narrower change is worth naming separately: a `json_format: array` dataset over a **zero-byte or whitespace-only** file now errors instead of returning zero rows. That matches what the buffered path already does for an explicitly-declared array, and an empty body cannot be distinguished from one truncated to nothing. `json_format: auto` and `json_format: json` are unaffected — they take the buffered branch, where an empty file still reads as zero rows.

## Test plan

- `make lint`
- `make test`
- New coverage in `crates/dataformat-json/src/stream.rs`: the issue's exact repro; every syntax-error shape (missing colon, bad token, trailing comma in object and in array, missing separator); truncation at a numeric boundary; scalars split across pushes; non-array bodies; BOM and leading whitespace byte-by-byte.
- Two round-trip tests split bodies at **every byte offset** — through escapes, multi-byte UTF-8, numbers and bare scalars — asserting identical rows and no error at any boundary. This is what caught change 2: without it the fix would have hard-errored on valid data.
- Five tests in `crates/dataformat-json/src/file_format.rs` cover the end-of-input transition through the real `deserialize_stream` wiring, not `flush` alone: an unclosed array, a non-array body, a closed array flushed repeatedly (the over-rejection guard), and a parity test asserting the buffered and streaming readers now reject the same body.
- Second neuter matrix over that fix, four arms, each isolating exactly its tests: drop the `finish` call → the 4 error tests fail; make `finish` always succeed → the same 4; make it reject `Complete` too → only the closed-array test; swap the two state messages → only the 2 message tests. Test counts summed to 381 on every arm, confirming no test was lost.
- Trailing-content coverage: `[1]garbage`, `[1][2]`, `[{"a":1}] {"b":2}`, `[{"a":1}]garbage`, `[1]]`, `[1],`, `[]x`; content arriving in a *later* push than the one that closed the array; and an over-rejection guard asserting `[1]\n`, `[1] `, `[{"a":1}]  `, CRLF/tab tails and `[]\n` still read normally and finish cleanly.
- Third neuter matrix, five arms, each isolating cleanly and every arm summing to the full test count: make the trailing check a no-op → the 2 trailing tests fail; remove it from `finish` alone → only the later-push test; remove it from the post-completion early return alone → only the later-push test; make it reject whitespace too → the over-rejection test **plus 2 pre-existing tests** (`test_push_whitespace_handling`, `flush_accepts_a_closed_array`); drop the tail discard → only the whitespace-tail test.
- Neuter matrix, one arm per direction of wrongness, each isolating cleanly: restore the syntax swallow → 4 malformation tests fail, 0 truncation tests; restore the array-start collapse → 2 non-array tests; drop the truncation guard → 2 truncation tests; drop the scalar guard → 2 scalar tests.

## Checklist

- [x] Regression test fails on the old code and passes on the new
- [x] Edge cases: empty, truncated, byte-by-byte, multi-byte UTF-8, escapes, nested, scalars, BOM
- [x] Scoped clippy over the crate and its tests — 0 warnings
- [x] No `.unwrap()` in added code
- [ ] `make signoff` / `Attestation`

Filed while widening this PR, deliberately **not** fixed here: #12782 — the pull-based reader (`ArrayToNdjson`) disagrees with the push adapter in both directions (it accepts `[{"a":1}]garbage`, returns a row of `1]` for `[1]]`) and separately cannot read `[1,2,3]` at all. Different reader, different mechanism.

Fixes #12755
Fixes #12780

### CODE REVIEW

**Comment by claudespice** [comment]
@copilot review

File: crates/dataformat-json/src/stream.rs [hunk 1]
```diff
@@ -972,15 +972,20 @@ impl ArrayToNdjsonPush {
...
+                    }
                 }
-            } else {
-                // Not enough data yet
-                return Ok(());
+                // Only a buffer that has not reached the `[` yet is waiting on
+                // more bytes. Any other error means the payload is not a JSON
+                // array at all — an NDJSON or object body read as one is a
+                // configuration mistake, not a short read.
+                Err(e) if e.kind() == io::ErrorKind::UnexpectedEof => return Ok(()),
```
[the remark this answers is not in the document]
Comment (claudespice) [reply crates/dataformat-json/src/stream.rs:1021]: Confirmed and fixed in 65926b3b0.

I measured the premise before acting, and it holds — with a sharper failure than the empty-body case. Feeding the adapter `[{"a":1},{"a":2}` (no closing bracket) emits **both rows, reports `is_complete() == false`, and returns no error**, because nothing ever asks it whether the array closed. So a file cut short by a partial fetch or an interrupted upload reads as a complete, shorter table — the same silent-row-loss class this PR exists to close, and one my own truncation guard here made slightly worse (`[1,2,3` now holds the trailing `3` back waiting for bytes that never arrive).

The buffered reader over the identical body already rejects it, so the two paths disagreed:

| body | `ArrayToNdjson` (buffered) | `ArrayToNdjsonPush` (streaming) |
| --- | --- | --- |
| `[{"a":1},{"a":2}` | `Err(UnexpectedEof)` | Ok — 2 rows, silent |
| `[1,2,3` | `Err(InvalidData)` | Ok — 2 rows, silent |
| `   ` | `Err(UnexpectedEof)` | Ok — 0 rows, silent |

Added `ArrayToNdjsonPush::finish`, which rejects every state but `Complete`, and called it from `SpiceJsonDecoder::flush` as you suggested. `flush` is the correct hook: `can_flush_early()` is false, and `DecoderDeserializer::next` only flushes when `decode` returns 0, which happens solely for the empty sentinel `finish()` pushes once the input is exhausted — so it cannot fire mid-stream.

I also checked the two ways a valid file could be mistaken for a truncated one, and neither applies: `Format::Array` is in `needs_non_repartitioned`, so a file is never range-split into fragments, and the decoder is constructed inside the per-file `open()` future, so state never carries across files.

One deliberate behaviour change worth naming: a `format: array` dataset over a zero-byte or whitespace-only file now errors instead of returning zero rows. That matches what the buffered path already does for an explicitly-declared array, and an empty body is indistinguishable from one truncated to nothing. `format: auto` and `format: json` are unaffected — they take the buffered branch, where an empty file still reads as zero rows.

Five tests added, each verified to fail without the fix via a four-arm neuter matrix (drop the `finish` call; make `finish` always succeed; make it reject `Complete` too; swap the two state messages) — every arm isolated exactly the intended tests.

File: crates/dataformat-json/src/stream.rs [hunk 2]
```diff
@@ -962,6 +962,40 @@ impl ArrayToNdjsonPush {
...
+    /// simply stops producing rows, and the short result is indistinguishable
+    /// from a complete one.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`io::ErrorKind::UnexpectedEof`] if the array was never opened
+    /// or never closed.
+    pub fn finish(&self) -> io::Result<()> {
+        let detail = match self.state {
+            ParsingState::Complete => return Ok(()),
```
[the remark this answers is not in the document]
Comment (claudespice) [reply crates/dataformat-json/src/stream.rs]: Confirmed and fixed in 7c5b1ea. Measured it before acting — `[1]garbage` and `[1][2]` both emitted the first array's rows, reported `is_complete() == true`, and returned `Ok` from `finish()`, exactly as you describe.

`ensure_only_trailing_whitespace` now runs at all three points where the state can be or become `Complete`: the closing-bracket transitions in `process_buffer`, the early return that swallowed later pushes, and `finish`. Whitespace is still accepted, so a file ending in a newline reads as it always did — `whitespace_after_the_closing_bracket_is_accepted` pins that, and neutering the check to reject whitespace as well fails it along with two pre-existing tests.

Regression tests added for `[1]garbage`, `[1][2]`, `[{"a":1}] {"b":2}`, `[{"a":1}]garbage`, `[1]]`, `[1],`, `[]x`, plus the case where the trailing bytes arrive in a *later* push than the one that closed the array — that one is only caught by the early-return path, and it is the arm that fails when either the `process_buffer` guard or the `finish` guard is removed on its own.

One thing your comment led me to that is not in this PR: the same check exposed that the *pull* reader (`ArrayToNdjson`) disagrees with the push adapter in both directions — it accepts `[{"a":1}]garbage`, returns a row of `1]` for `[1]]`, and separately cannot read `[1,2,3]` at all. That is a different reader with a different mechanism, so it is filed as #12782 rather than widened into this PR.

Follow-up in 69afe92: bytes after the array closes were accumulated but never parsed, so checking them re-scanned the whole tail on every push. Verified whitespace is now dropped instead.

**Comment by claudespice** [comment]
The red `Attestation` here is **not this diff**. Re-dispatched the sign-off once; no code change.

Run [`31239284344`](https://github.com/spiceai/spiceai/actions/runs/31239284344) reached a real verdict at 324.8 min (it did not wedge and did not expire its budget) and failed on exactly one test:

```
Summary [2818.304s] 9972 tests run: 9971 passed (14 slow), 1 timed out, 27 skipped
TIMEOUT [ 960.038s] (6657/9972) cayenne::mutation_property_test prop_concurrent_mixed_position_sqlite
```

This branch touches `crates/dataformat-json/src/file_format.rs` and `crates/dataformat-json/src/stream.rs` — nothing that can reach `crates/cayenne`. The failing test was killed at the wall-clock ceiling, with no assertion, panic or compile error anywhere in the log.

The same run's matched siblings show the box was heavily contended: `prop_concurrent_upsert_only_key_sqlite` **passed at 422.070s** against a 92.0s quiet-pool baseline (4.59x), and `prop_concurrent_upsert_only_position_sqlite` passed at 514.657s. Because that binary is pinned to `retries = 0`, one contention timeout is an unrecoverable sign-off failure.

Filed #12811 for the ceiling itself — it was sized against a 3.7x worst-case contention factor that this run exceeded. That is a CI-config question, not something to fix in this PR.

New sign-off: [`31252334930`](https://github.com/spiceai/spiceai/actions/runs/31252334930). This is the first retry — the branch's three earlier runs were all `cancelled` concurrency evictions, so `31239284344` is the only verdict this head has received.

**Review by phillipleblanc — APPROVED** [review]
